src/server/scripts/Northrend/Naxxramas/boss_thaddius.cpp

```diff
@@ -110,168 +155,302 @@ class boss_thaddius : public CreatureScript
+                if (Creature* feugen = ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_FEUGEN)))
+                    if (feugen->AI())
+                        feugen->AI()->DoAction(ACTION_BEGIN_RESET_ENCOUNTER);
+                if (Creature* stalagg = ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_STALAGG)))
+                    if (stalagg->AI())
+                        stalagg->AI()->DoAction(ACTION_BEGIN_RESET_ENCOUNTER);
+
+                me->setActive(false);
+            }
 
-            if (checkStalaggAlive != checkFeugenAlive)
+            void ResetEncounter()
             {
-                uiAddsTimer += diff;
-                if (uiAddsTimer > 5000)
-                {
-                    if (!checkStalaggAlive)
-                    {
-                        if (Creature* pStalagg = ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_STALAGG)))
-                            pStalagg->Respawn();
-                    }
-                    else
-                    {
-                        if (Creature* pFeugen = ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_FEUGEN)))
-                            pFeugen->Respawn();
-                    }
-                }
+                events.SetPhase(PHASE_NOT_ENGAGED);
+                _phase = PHASE_NOT_ENGAGED;
+                feugenAlive = true;
+                stalaggAlive = true;
+                me->Respawn(true);
+                me->CastSpell(me, SPELL_THADDIUS_INACTIVE_VISUAL);
+
+                if (Creature* feugen = ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_FEUGEN)))
+                    if (feugen->AI())
+                        feugen->AI()->DoAction(ACTION_RESET_ENCOUNTER);
+                if (Creature* stalagg = ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_STALAGG)))
+                    if (stalagg->AI())
+                        stalagg->AI()->DoAction(ACTION_RESET_ENCOUNTER);
+                _Reset();
             }
 
-            if (!UpdateVictim())
-                return;
-
-            events.Update(diff);
-
-            if (me->HasUnitState(UNIT_STATE_CASTING))
-                return;
-
-            while (uint32 eventId = events.ExecuteEvent())
+            void UpdateAI(uint32 diff) override
             {
-                switch (eventId)
+                switch (_phase)
                 {
-                    case EVENT_SHIFT:
-                        DoCastAOE(SPELL_POLARITY_SHIFT);
-                        events.ScheduleEvent(EVENT_SHIFT, 30000);
-                        return;
-                    case EVENT_CHAIN:
-                        DoCastVictim(SPELL_CHAIN_LIGHTNING);
-                        events.ScheduleEvent(EVENT_CHAIN, urand(10000, 20000));
-                        return;
-                    case EVENT_BERSERK:
-                        DoCast(me, SPELL_BERSERK);
+                    case PHASE_NOT_ENGAGED:
                         return;
```

Comment:
No events should happen while we're not in combat, so we save ourselves a few cycles and just return right away. 

PHASE_NOT_ENGAGED only happens right after init (no events in the queue) or after a ResetEncounter() call. ResetEncounter() calls BossAI::_Reset(), which clears all events.